zookeeper/core/field.py

```diff
@@ -0,0 +1,230 @@
+import inspect
+from typing import Callable, Generic, Type, TypeVar, Union
+
+from zookeeper.core import utils
+from zookeeper.core.partial_component import PartialComponent
+
+
+# A sentinel class/object for missing default values.
+class _MISSING:
+    def __repr__(self):
+        return f"<missing>"
+
+
+_missing = _MISSING()
+
+
+# Type-variables to parameterise fields.
+_ComponentType = TypeVar("_ComponentType")
+_FieldType = TypeVar("_FieldType")
+
+
+class Field(Generic[_ComponentType, _FieldType]):
+    """
+    A configurable field for Zookeeper components. Fields must be typed, may
+    take default values, and are configurable through the CLI.
+
+    This class is not appropriate for nesting child sub-components; for this,
+    use `ComponentField` instead.
+
+    Internally, it is expected that each component will cache the result of
+    `get_default`.
+    """
+
+    def __init__(
+        self,
+        default_value: Union[
+            _MISSING,
+            _FieldType,
+            Callable[[], _FieldType],
+            Callable[[_ComponentType], _FieldType],
+        ] = _missing,
+    ):
+        # Define here once to avoid having to define twice below.
+        default_value_error = TypeError(
+            "If `default_value` is passed to `Field`, it must be either:\n"
+            "- An immutable value (int, float, bool, string, or None).\n"
+            "- A function or lambda accepting no arguments or a single\n"
+            "  argument (`self`), and returning the default value.\n"
+            f"Received: {default_value}."
+        )
+
+        return_annotation = inspect.Signature.empty
+
+        if default_value is _missing:
+            default_factory = None
+        elif isinstance(default_value, (int, float, bool, str, type(None))):
+            default_factory = lambda instance: default_value  # noqa: E731
+        elif inspect.isfunction(default_value):
+            signature = inspect.signature(default_value)
+            return_annotation = signature.return_annotation
+            if len(signature.parameters) == 0:
+                default_factory = lambda instance: default_value()  # noqa: E731
+            elif len(signature.parameters) == 1 and all(
+                signature.parameters[name].kind
+                not in (
+                    inspect.Parameter.VAR_POSITIONAL,
+                    inspect.Parameter.VAR_KEYWORD,
+                )
+                for name in signature.parameters
+            ):
+                default_factory = default_value
+            else:
+                raise default_value_error
+        else:
+            raise default_value_error
+
+        self._registered = False
+        self._return_annotation = return_annotation
+        self._default_factory = default_factory
```

Comment:
I think I would prefer doing those checks in `get_default`, that would have the benefit that we don't need to wrap the values in a factory method, though unfortunately it would also get rid of the early error message. Do you think there is an easy way to get around this?

Reply:
You mean then we wouldn't have to wrap the immutable values in a lambda? Well, we can just not wrap immutables and check `inspect.isfunction` in `get_default`. I don't mind doing that, although yes it would get rid of early warning messages.